data_steward/analytics/cdr_ops/rdr_export_qc.py

```diff
@@ -0,0 +1,308 @@
+    """
+    Raise an error if a required parameter is missing from the params dict
+    """
+    for required_param in required_params:
+        if required_param not in params:
+            raise ValueError(f'Missing query parameter `{required_param}`')
+    global project_id, old_rdr, new_rdr, cope_version_map
+    project_id = params.get('project_id')
+    old_rdr = params.get('old_rdr')
+    new_rdr = params.get('new_rdr')
+    print(f'All required parameters are specified.\n{params}')
+
+
+# -
+
+params = get_url_params(NOTEBOOK_URL)
+validate_params(params, REQUIRED_PARAMS)
+
+# # Missing tables
+
+query = f'''
+SELECT 
+ COALESCE(curr.table_id, prev.table_id) AS table_id, 
+ curr.row_count AS {new_rdr}, 
+ prev.row_count AS {old_rdr}, 
+ (curr.row_count - prev.row_count) row_diff 
+FROM {project_id}.{new_rdr}.__TABLES__ curr 
+FULL OUTER JOIN {old_rdr}.__TABLES__ prev
+  USING (table_id)
+WHERE curr.table_id IS NULL OR prev.table_id IS NULL
+'''
+pd.read_gbq(query, dialect='standard')
+
+# ## Row count comparison
+
+query = f'''
+SELECT 
+ curr.table_id AS table_id, 
+ curr.row_count AS {new_rdr}, 
+ prev.row_count AS {old_rdr}, 
+ (curr.row_count - prev.row_count) row_diff 
+FROM {project_id}.{new_rdr}.__TABLES__ curr 
+JOIN {project_id}.{old_rdr}.__TABLES__ prev
+  USING (table_id)
+ORDER BY ABS(curr.row_count - prev.row_count) DESC;
+'''
+pd.read_gbq(query, dialect='standard')
+
+# ## Concept codes used
+# Identify question and answer concept codes which were either added or removed (appear in only the new or only the old RDR datasets, respectively).
+
+query = f'''
+WITH curr_code AS (
+SELECT observation_source_value value, 
+ 'observation_source_value' field, 
+ COUNT(1) row_count 
+FROM {project_id}.{new_rdr}.observation GROUP BY 1
+
+UNION ALL
+
+SELECT value_source_value value, 
+ 'value_source_value' field, 
+ COUNT(1) row_count 
+FROM {project_id}.{new_rdr}.observation GROUP BY 1),
+
+prev_code AS (
+SELECT observation_source_value value,
+ 'observation_source_value' field, 
+ COUNT(1) row_count 
+FROM {project_id}.{old_rdr}.observation GROUP BY 1
+
+UNION ALL
+
+SELECT value_source_value value, 
+ 'value_source_value' field, 
+ COUNT(1) row_count 
+FROM {project_id}.{old_rdr}.observation GROUP BY 1)
+
+SELECT 
+  prev_code.value prev_code_value, 
+  prev_code.field prev_code_field, 
+  prev_code.row_count prev_code_row_count, 
+  curr_code.value curr_code_value, 
+  curr_code.field curr_code_field, 
+  curr_code.row_count curr_code_row_count, 
+FROM curr_code 
+ FULL OUTER JOIN prev_code
+  USING (field, value)
+WHERE prev_code.value IS NULL OR curr_code.value IS NULL
+'''
+pd.read_gbq(query, dialect='standard')
+
+# # Check if observation_source_value vs concept ids
+
+query = f"""
+SELECT 
+    SUM(CASE WHEN observation_source_concept_id IS NULL THEN 1 ELSE 0 END) as n_null_source_concept_id
+    , SUM(CASE WHEN observation_source_concept_id=0 THEN 1 ELSE 0 END) as n_zero_source_concept_id
```

Comment:
opt:  just a personal preference.  you can leave as is, if you want.  i normally expect the comma to come at the end of the line instead of the start of the next line.

Reply:
This is a style preference I'm somewhat conflicted about. In a previous life my team would **always** structure queries like this
```sql
SELECT
  col1
 ,col2
 ,col3
FROM `project.dataset.table`
WHERE 1=1    -- wat! yes, this conveniently ensures each condition in the clause is prefixed with "AND ..."
 AND col1 < col3
 AND col2 > col4
```
because it makes it easier to remove items from the selection and the where clause. These are things one typically does when iterating on a query or troubleshooting data issues.

I might be willing to switch all notebook queries to this style- or allow comma to be in _either_ the end or beginning of the line. Thoughts?

The following style is rarely desirable and I think we should discourage it:
```
SELECT col1, col2, col3 FROM `project.dataset.table` WHERE col1 < col3 AND col2 > col4
```

Reply:
If you want to leave the commas at the start of the line, I can support that.  I guess I'm nitpicking about alignment.  I would prefer to keep the alphabetic characters lined up like in your example query above.  So, maybe:
```
     SUM.......
    ,SUM.......
```
instead of:
```
    SUM.....
    , SUM.....
```

Reply:
Ok I aligned the queries as you mentioned above